GH-38348: [C#] Make PrimitiveArray<T> support IReadOnlyList<T?>

voidstar69 opened the pull request:
### What changes are included in this PR?

Make Arrow arrays of scalar type T implement the same semantic contract as IReadOnlyList<T?>.
Note that this PR does not include similar support for ICollection<T?>. I could add that support in this PR or a future PR.

### Are these changes tested?

This PR includes unit tests of the implemented IReadOnlyList<T?> methods.
* Closes: #38348

CurtHagenlocher commented:
I was thinking this would include all scalar arrays, e.g. TimestampArray : IReadOnlyList<DateTime?>, StringArray : IReadOnlyList<string>, TimeXXArray : IReadOnlyList<TimeOnly?>, DateXXArray : IReadOnlyList<DateOnly?>, IReadOnlyList<DateTime?>, BooleanArray : IReadOnlyList<bool?>, BinaryArray : IReadOnlyList<byte[]>.

The original issue also suggests implementing ICollection<T>.

Would you file followup work items so that this one can be marked as closed if checked-in as-is?

voidstar69 commented:
> I was thinking this would include all scalar arrays, e.g. TimestampArray : IReadOnlyList<DateTime?>, StringArray : IReadOnlyList, TimeXXArray : IReadOnlyList<TimeOnly?>, DateXXArray : IReadOnlyList<DateOnly?>, IReadOnlyList<DateTime?>, BooleanArray : IReadOnlyList<bool?>, BinaryArray : IReadOnlyList<byte[]>.
> 
> The original issue also suggests implementing ICollection.
> 
> Would you file followup work items so that this one can be marked as closed if checked-in as-is?

I have implemented this for all the scalar arrays in your example. Are there any more scalar arrays that need to support `IReadOnlyList`?

For `TimestampArray` it seemed more natural for it to implement `IReadOnlyList<DateTimeOffset?>`.

Note that I implemented the various `GetEnumerator` methods in two distinct ways, a verbose for-loop and a concise LINQ statement. Let me know which you think is better and I will standardise on that form.

I created a followup issue #38692 for adding support for `ICollection<T?>`.

CurtHagenlocher commented:
> Note that I implemented the various `GetEnumerator` methods in two distinct ways, a verbose for-loop and a concise LINQ statement. Let me know which you think is better and I will standardise on that form.

While concision is very nice, a for-loop allows inlining in a way that delegates don't (or don't yet at least) and I think that performance is more important than concision code in infrastructure.
